src/provider/model/FinEnvironment.ts

```diff
@@ -114,48 +114,91 @@ export class FinEnvironment extends AsyncInit implements Environment {
             await Injector.initialized;
+            this.deregisterApplication({uuid: event.uuid});
+        };
+
+        fin.System.addListener('application-closed', appicationClosedHandler);
+        fin.System.addListener('application-crashed', appicationClosedHandler);
+
+        fin.System.addListener('window-created', async (event: WindowEvent<'system', 'window-created'>) => {
             const identity = {uuid: event.uuid, name: event.name};
 
-            this._createdWindows.delete(getId(identity));
+            await Injector.initialized;
+            this.registerApplication({uuid: event.uuid}, Promise.resolve());
+            this.registerWindow(identity);
+        });
+
+        fin.System.addListener('window-closed', async (event: WindowEvent<'system', 'window-closed'>) => {
+            const identity = {uuid: event.uuid, name: event.name};
 
-            this.windowClosed.emit(identity);
+            await Injector.initialized;
+            this.deregisterWindow(identity);
+            this.deregisterApplication({uuid: event.uuid});
         });
 
         // No await here otherwise the injector will never properly initialize - The injector awaits this init before completion!
         Injector.initialized.then(async () => {
+            // Register windows that were running before launching the FDC3 service
             windowInfo.forEach(info => {
                 const {uuid, mainWindow, childWindows} = info;
+                this.registerApplication({uuid: info.uuid}, undefined);
 
-                this.registerWindow({uuid, name: mainWindow.name}, undefined);
-                childWindows.forEach(child => this.registerWindow({uuid, name: child.name}, undefined));
+                this.registerWindow({uuid, name: mainWindow.name});
+                childWindows.forEach(child => this.registerWindow({uuid, name: child.name}));
             });
         });
     }
 
-    private async registerWindow(identity: Identity, creationTime: number | undefined): Promise<void> {
-        const createdWindow = {
-            creationTime,
-            index: this._windowsCreated
-        };
+    private registerWindow(identity: Identity): void {
+        if (identity.uuid !== SERVICE_IDENTITY.uuid) {
+            const createdWindow = {index: this._windowsCreated};
 
-        this._createdWindows.set(getId(identity), createdWindow);
-        this._windowsCreated++;
+            this._windows.set(getId(identity), createdWindow);
+            this._windowsCreated++;
 
-        this.windowCreated.emit(identity);
+            this.windowCreated.emit(identity);
+        }
+    }
+
+    private deregisterWindow(identity: Identity): void {
+        if (identity.uuid !== SERVICE_IDENTITY.uuid) {
+            this._windows.delete(getId(identity));
+            this.windowClosed.emit(identity);
+        }
+    }
+
+    private registerApplication(identity: Identity, startedPromise: Promise<void> | undefined): void {
```

Comment:
I feel like using **?** is easier to read/notice 
`startedPromise?: Promise<void>`

Reply:
I'm against doing this, as I want the caller to *know* they're passing in undefined here - in this context, undefined should be a deliberate choice, rather than a fallback